### PR TITLE
Try catch around add app setting

### PR DESCRIPTION
**Pull request by nturinski**
Fixes https://github.com/Microsoft/vscode-azureappservice/issues/943

### CODE REVIEW

File: appservice/src/tree/AppSettingsTreeItem.ts [hunk 1]
```diff
@@ -109,7 +109,15 @@ export class AppSettingsTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
 
         showCreatingTreeItem(newKey);
         settings.properties[newKey] = newValue;
-        await this.root.client.updateApplicationSettings(settings);
+
+        try {
+            await this.root.client.updateApplicationSettings(settings);
+        } catch (err) {
+            // if this is rejected, we need to clear the property otherwise it will cause every subsequent add to fail
+            delete settings.properties[newKey];
```
Comment (EricJizbaMSFT) [inline appservice/src/tree/AppSettingsTreeItem.ts]: I'm guessing `editSettingItem` (which is used for "edit" and "rename" of an already existing setting) will have a similar problem

Comment (nturinski) [reply appservice/src/tree/AppSettingsTreeItem.ts]: Good point

**Comment by nturinski** [comment]
Just a note, I'm not very worried about edit value failing since this passes:

image

Also I'm calling `editSettingsValue` first now because if it failed out, the name value would be wrong until refresh (because the reassign happens before the error is thrown)

File: appservice/src/tree/AppSettingsTreeItem.ts [hunk 2]
```diff
@@ -71,14 +71,24 @@ export class AppSettingsTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
     public async editSettingItem(oldKey: string, newKey: string, value: string): Promise<void> {
         const settings: StringDictionary = await this.ensureSettings();
```
Comment (EricJizbaMSFT) [inline appservice/src/tree/AppSettingsTreeItem.ts]:
```
// clone cached settings into new object
const settings: StringDictionary = { ...await this.ensureSettings() };
```

File: appservice/src/tree/AppSettingsTreeItem.ts [hunk 3]
```diff
@@ -71,14 +71,24 @@ export class AppSettingsTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
...
+
+        try {
+            await this.root.client.updateApplicationSettings(settings);
+        } catch (err) {
+            // if this was a rename, delete the invalid key and readd the old key
+            renameHelper(newKey, oldKey);
+            throw err;
         }
 
-        await this.root.client.updateApplicationSettings(settings);
```
Comment (EricJizbaMSFT) [inline appservice/src/tree/AppSettingsTreeItem.ts]: this._settings = await this.root.client.updateApplicationSettings(settings);

File: appservice/src/tree/AppSettingsTreeItem.ts (hunk 2 above)
Comment (nturinski) [reply appservice/src/tree/AppSettingsTreeItem.ts]: This is a great idea, but apparently only makes a shallow copy so anytime properties were referenced, it was still referring to the `this._settings.properties`.  I used the JSON method to make a deep-copy

Ref: https://scotch.io/bar-talk/copying-objects-in-javascript

File: appservice/src/tree/AppSettingsTreeItem.ts [hunk 4]
```diff
@@ -69,16 +69,17 @@ export class AppSettingsTreeItem extends AzureParentTreeItem<ISiteTreeRoot> {
...
                 delete settings.properties[oldKey];
             }
             settings.properties[newKey] = value;
         }
 
-        await this.root.client.updateApplicationSettings(settings);
+        this._settings = await this.root.client.updateApplicationSettings(settings);
     }
 
     public async deleteSettingItem(key: string): Promise<void> {
```
Comment (EricJizbaMSFT) [inline appservice/src/tree/AppSettingsTreeItem.ts]: Seems like you might as well do the same thing for delete, too

**Pull request merged**